### PR TITLE
feat: add `IStaticWeightedIsm` implementation

### PR DESCRIPTION
**Pull request by aroralanuk**
### Description

- Adding the static variation of the IWeightedIsm config where you configure validators with their corresponding weights and store it as bytecode.

### Drive-by changes

None

### Related issues

- fixes https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/4160

### Backward compatibility

Yes

### Testing

Fuzz tests

### CODE REVIEW

File: solidity/contracts/isms/multisig/AbstractMultisigIsm.sol [hunk 1]
```diff
@@ -57,6 +52,38 @@ abstract contract AbstractMultisigIsm is IMultisigIsm {
...
+        bytes calldata _metadata
+    ) public pure virtual returns (uint256);
+}
+
+/**
+ * @title AbstractMultisigIsm
+ * @notice Manages per-domain m-of-n Validator sets of AbstractMultisig that are used to verify
+ * interchain messages.
+ */
+abstract contract AbstractMultisigIsm is AbstractMultisig {
```
Comment (yorhodes) [inline solidity/contracts/isms/multisig/AbstractMultisigIsm.sol:70]: oh I see

File: solidity/contracts/isms/multisig/AbstractMultisigIsm.sol
[the remark this answers is not in the document]
Comment (yorhodes) [reply solidity/contracts/isms/multisig/AbstractMultisigIsm.sol:1]: nvm, I see you just made an intermediate abstraction

File: solidity/contracts/isms/multisig/AbstractMultisigIsm.sol (hunk 1 above)
Comment (yorhodes) [inline solidity/contracts/isms/multisig/AbstractMultisigIsm.sol:70]: should we make an `AbstractWeightedMultisigIsm`?

File: solidity/contracts/isms/multisig/AbstractWeightedMultisigIsm.sol [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+ * @title AbstractStaticWeightedMultisigIsm
+ * @notice Manages per-domain m-of-n Validator sets with stake weights that are used to verify
+ * interchain messages.
+ */
+abstract contract AbstractStaticWeightedMultisigIsm is
+    AbstractMultisig,
+    IStaticWeightedMultisigIsm
+{
+    // ============ Constants ============
+    uint96 public constant BASIS_POINTS = 10000;
```
Comment (yorhodes) [inline solidity/contracts/isms/multisig/AbstractWeightedMultisigIsm.sol:38]: im not sure what this means

Comment (yorhodes) [reply solidity/contracts/isms/multisig/AbstractWeightedMultisigIsm.sol:38]: oh now I see that you are treating this as "percentages"
I would personally just use absolute stake amounts as weights for simplicity

File: solidity/contracts/isms/multisig/AbstractWeightedMultisigIsm.sol [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+                _signer != _validators[_validatorIndex].signingAddress
+            ) {
+                ++_validatorIndex;
+            }
+            // fail if we never found a match
+            require(_validatorIndex < _validatorCount, "Invalid signer");
+
+            // add the weight of the current validator
+            _totalWeight += _validators[_validatorIndex].weight;
+        }
```
Comment (yorhodes) [inline solidity/contracts/isms/multisig/AbstractWeightedMultisigIsm.sol:91]: unfortunate that we have to duplicate some of this logic but I dont see an easy way to reuse

File: solidity/contracts/isms/multisig/WeightedMultisigIsm.sol [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+import {IInterchainSecurityModule} from "../../interfaces/IInterchainSecurityModule.sol";
+import {AbstractMerkleRootMultisigIsm} from "./AbstractMerkleRootMultisigIsm.sol";
+import {AbstractMessageIdMultisigIsm} from "./AbstractMessageIdMultisigIsm.sol";
+import {AbstractStaticWeightedMultisigIsm} from "./AbstractWeightedMultisigIsm.sol";
+import {AbstractMultisigIsm} from "./AbstractMultisigIsm.sol";
+import {StaticWeightedValidatorSetFactory} from "../../libs/StaticWeightedValidatorSetFactory.sol";
+import {MetaProxy} from "../../libs/MetaProxy.sol";
+
+abstract contract AbstractMetaProxyMultisigIsm is
```
Comment (yorhodes) [inline solidity/contracts/isms/multisig/WeightedMultisigIsm.sol:24]: this is overloaded with the the ones in `AbstractMultisigIsm`

**Review by yorhodes — APPROVED** [review]
this is looking great
nice job reusing the existing abstractions
would be cool in future to have a threshold weight that adjusts with the message content

File: solidity/contracts/interfaces/isms/IWeightedMultisigIsm.sol [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+import {IInterchainSecurityModule} from "../IInterchainSecurityModule.sol";
+
+interface IStaticWeightedMultisigIsm is IInterchainSecurityModule {
+    // ============ Structs ============
+
+    // ValidatorInfo contains the signing address and weight of a validator
+    struct ValidatorInfo {
+        address signingAddress;
+        uint96 weight;
+    }
```
[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/contracts/interfaces/isms/IWeightedMultisigIsm.sol:25]: Yes, but I want to use the same struct for the mutable non-bytecode one too.

File: solidity/contracts/interfaces/isms/IWeightedMultisigIsm.sol [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+     * @param _message The message to be verified
+     * @return validators The validators and their weights
+     * @return thresholdWeight The threshold weight required to pass verification
+     */
+    function validatorsAndThresholdWeight(
+        bytes calldata _message
+    )
+        external
+        view
+        returns (ValidatorInfo[] memory validators, uint96 thresholdWeight);
```
[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/contracts/interfaces/isms/IWeightedMultisigIsm.sol:38]: for storing the validator in storage.

File: solidity/contracts/isms/multisig/AbstractMerkleRootMultisigIsm.sol [hunk 7]
```diff
@@ -61,12 +57,21 @@ abstract contract AbstractMerkleRootMultisigIsm is AbstractMultisigIsm {
...
         bytes calldata _metadata,
         uint256 _index
     ) internal pure virtual override returns (bytes calldata) {
         return _metadata.signatureAt(_index);
     }
+
+    /**
+     * @inheritdoc AbstractMultisig
+     */
+    function signatureCount(
```
[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/contracts/isms/multisig/AbstractMerkleRootMultisigIsm.sol:72]: I mentioned it in our discussion, for making sure we don't pass in inadequate number of signatures.

File: solidity/contracts/isms/multisig/AbstractMessageIdMultisigIsm.sol [hunk 8]
```diff
@@ -46,12 +42,21 @@ abstract contract AbstractMessageIdMultisigIsm is AbstractMultisigIsm {
...
         bytes calldata _metadata,
         uint256 _index
     ) internal pure virtual override returns (bytes calldata) {
         return _metadata.signatureAt(_index);
     }
+
+    /**
+     * @inheritdoc AbstractMultisig
+     */
+    function signatureCount(
```
[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/contracts/isms/multisig/AbstractMessageIdMultisigIsm.sol:57]: same as above

File: solidity/contracts/isms/multisig/AbstractMultisigIsm.sol (hunk 1 above)
Comment (aroralanuk) [reply solidity/contracts/isms/multisig/AbstractMultisigIsm.sol:70]: That's the non-static one.

File: solidity/contracts/isms/multisig/AbstractWeightedMultisigIsm.sol (hunk 2 above)
Comment (aroralanuk) [reply solidity/contracts/isms/multisig/AbstractWeightedMultisigIsm.sol:38]: Then we'll need to keep updating the threshold weight.

Comment (aroralanuk) [reply solidity/contracts/isms/multisig/AbstractWeightedMultisigIsm.sol:38]: renamed to TOTAL_WEIGHT and using 1e10 instead.

Comment (yorhodes) [reply solidity/contracts/isms/multisig/AbstractWeightedMultisigIsm.sol]: > Then we'll need to keep updating the threshold weight.
I still view the "updates" as relatively infrequent (on an epoch/monthly basis)

File: solidity/contracts/isms/multisig/AbstractMerkleRootMultisigIsm.sol (hunk 7 above)
[the remark this answers is not in the document]
Comment (yorhodes) [reply solidity/contracts/isms/multisig/AbstractMerkleRootMultisigIsm.sol:72]: the relayer would not do a view call to check this so a view function does not make sense

File: solidity/contracts/interfaces/isms/IWeightedMultisigIsm.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (yorhodes) [reply solidity/contracts/interfaces/isms/IWeightedMultisigIsm.sol:25]: imo we would not want to use that one
we dont have a mutable multisig ISM so why bias toward this use case now?

File: solidity/contracts/isms/multisig/AbstractMerkleRootMultisigIsm.sol (hunk 7 above)
[the remark this answers is not in the document]
Comment (yorhodes) [reply solidity/contracts/isms/multisig/AbstractMerkleRootMultisigIsm.sol:72]: adding additional public/external functions unnecessarily pollutes the jump table, imposing additional gas overhead on calls (every ism.verify)

**Pull request merged**